### PR TITLE
Add openapi spec

### LINKED ISSUE #1503: Rate limiting of v1beta1.external.metrics.k8s.io
I have keda deployed in all of my environments and I notice that I am getting spammed in all of them with rate limiting notices. Including environments which are not actually using Keda at all yet. Is there anything that I should be doing to address this, or is this a known issue of some sort? I don't see any documentation around this situation. I haven't seen any ill effect on keda's performance where I am actually using it.

```
kube-apiserver-ip-10-28-11-137.us-west-1.compute.internal kube-apiserver E0112 17:07:58.690813       1 controller.go:114] loading OpenAPI spec for "v1beta1.external.metrics.k8s.io" failed with: OpenAPI spec does not exist
kube-apiserver-ip-10-28-11-137.us-west-1.compute.internal kube-apiserver I0112 17:07:58.690830       1 controller.go:127] OpenAPI AggregationController: action for item v1beta1.external.metrics.k8s.io: Rate Limited Requeue.
```

Keda version: 2.0.0 
Kubernetes Version: 1.18.6

### PR DESCRIPTION
**Pull request by surki**
Signed-off-by: Suresh Kumar Ponnusamy <suresh.ponnusamy@freshworks.com>

Add OpenAPI spec support, without which k8s apiserver log is flooded with errors/retries.
This follows similar upstream changes done [here](https://github.com/kubernetes-sigs/custom-metrics-apiserver/commit/edd77be65ba1e67c5091377be4b3ef864d52f50b)

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [x] Changelog has been updated

Fixes #1503

### CODE REVIEW

**Comment by surki** [comment]
@zroubalik  looks like we move `pkg/generated` this away, generate again and compare. Should I move `pkg/generated/openapi` into `adapter/generated` or somewhere else?

Or should I move the openapi spec generation logic into `hack/update-codegen.sh`? (not sure the role of `make generate`)

**Comment by zroubalik** [comment]
Thanks for doing this!

> @zroubalik looks like we move `pkg/generated` this away, generate again and compare. Should I move `pkg/generated/openapi` into `adapter/generated` or somewhere else?
> 
> Or should I move the openapi spec generation logic into `hack/update-codegen.sh`? (not sure the role of `make generate`)

I'd move it into `adapter/generated`.

File: adapter/main.go [hunk 1]
```diff
@@ -8,22 +8,23 @@ import (
 	"strconv"
 	"time"
 
+	kedav1alpha1 "github.com/kedacore/keda/v2/api/v1alpha1"
+	generatedopenapi "github.com/kedacore/keda/v2/pkg/generated/openapi"
+	prommetrics "github.com/kedacore/keda/v2/pkg/metrics"
+	kedaprovider "github.com/kedacore/keda/v2/pkg/provider"
+	"github.com/kedacore/keda/v2/pkg/scaling"
+	"github.com/kedacore/keda/v2/version"
```
Comment (zroubalik) [inline adapter/main.go:16]: Minor nit, could you please move KEDA related imports into a separate block (intended by one line) under the generic imports? As it was before?

File: adapter/main.go [hunk 2]
```diff
@@ -111,6 +112,11 @@ func main() {
 	printVersion()
 
 	cmd := &Adapter{}
+
+	cmd.OpenAPIConfig = genericapiserver.DefaultOpenAPIConfig(generatedopenapi.GetOpenAPIDefinitions, openapinamer.NewDefinitionNamer(scheme.Scheme))
+	cmd.OpenAPIConfig.Info.Title = "keda-adapter"
+	cmd.OpenAPIConfig.Info.Version = "1.0.0"
```
Comment (zroubalik) [inline adapter/main.go:118]: Should this version colerate with KEDA version or with OpenAPI spec version?

File: CHANGELOG.md [hunk 3]
```diff
@@ -39,6 +39,7 @@
 - Improve performance when fetching current scaling information on Deployments ([#1458](https://github.com/kedacore/keda/pull/1458))
 - Improve error reporting in prometheus scaler ([#1497](https://github.com/kedacore/keda/pull/1497))
 - Check that metricNames are unique in ScaledObject ([#1390](https://github.com/kedacore/keda/pull/1390))
+- Serve OpenAPI spec from keda metrics apiserver ([#1512](https://github.com/kedacore/keda/pull/1512))
```
Comment (zroubalik) [inline CHANGELOG.md:42]:
```suggestion
- Serve OpenAPI spec from KEDA Metrics Apiserver ([#1512](https://github.com/kedacore/keda/pull/1512))
```

**Review by zroubalik — CHANGES_REQUESTED** [review]

File: adapter/main.go (hunk 1 above)
Comment (surki) [reply adapter/main.go]: Sure, fixed them. I have `goimports` running on-file-save, it auto arranged the imports.

File: adapter/main.go (hunk 2 above)
Comment (surki) [reply adapter/main.go:120]: This is "application api version", not application version or spec version. Please see here: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#infoObject

**Comment by zroubalik** [comment]
@surki you can ignore the failing static check (I will address that in a separate PR), just please add exclude-rule for the `staticcheck` linter for the `pkg/scalers/stackdriver_client.go` in https://github.com/kedacore/keda/blob/main/.golangci.yml#L50 and please add there a comment with explanation and link to this PR.

File: adapter/main.go (hunk 2 above)
Comment (zroubalik) [reply adapter/main.go:120]: okay, thanks!

**Comment by surki** [comment]
@zroubalik looks like the `CI / Validate PR (pull_request) ` failure onis due to [this](https://github.com/kedacore/keda/blob/main/.github/workflows/pr-validation.yml#L11) . We appear to be using tools.go from `main` branch, so it is not pulling new dependency that is introduced in this PR.

**Comment by zroubalik** [comment]
> @zroubalik looks like the `CI / Validate PR (pull_request) ` failure is due to [this](https://github.com/kedacore/keda/blob/main/.github/workflows/pr-validation.yml#L11) . We appear to be using tools.go from `main` branch, so it is not pulling new dependency that is introduced in this PR.

@surki Sorry I am not sure I fully understand, that line in pr-validation.yml is about the container that's being used to do the build, you can see the Dockerfile for this container here: https://github.com/kedacore/keda/blob/main/tools/build-tools.Dockerfile

**Comment by surki** [comment]
> > @zroubalik looks like the `CI / Validate PR (pull_request) ` failure is due to [this](https://github.com/kedacore/keda/blob/main/.github/workflows/pr-validation.yml#L11) . We appear to be using tools.go from `main` branch, so it is not pulling new dependency that is introduced in this PR.
> 
> @surki Sorry I am not sure I fully understand, that line in pr-validation.yml is about the container that's being used to do the build, you can see the Dockerfile for this container here: https://github.com/kedacore/keda/blob/main/tools/build-tools.Dockerfile

The new dependency we have added here in [tools.go](https://github.com/surki/keda/blob/fix-openapi-spec-errors/hack/tools.go#L9) is not being pulled in (but looks like controller-gen etc are already pulled in as part of Docker image). 

I could think of two options:
* Add a step in workflow file to sync go modules. Something like
```
@@ -23,6 +23,9 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-go-
 
+      - name: Go modules sync
+        run: go mod tidy
+
       - name: Verify Generated clientset is up to date
         run: make clientset-verify
```
* Update the build-tools.Dockerfile to install openapi-gen

I would suggest to go with first option. are you okay if I do that?

**Comment by zroubalik** [comment]
@surki gotcha. Syncing the go.mod sounds like a good idea!

File: Makefile [hunk 4]
```diff
@@ -173,9 +173,20 @@ manifests: controller-gen
 
 # Generate code (API)
 .PHONY: generate
-generate: controller-gen
+generate: controller-gen adapter/generated/openapi/zz_generated.openapi.go
 	$(CONTROLLER_GEN) object:headerFile="hack/boilerplate.go.txt" paths="./..."
 
+adapter/generated/openapi/zz_generated.openapi.go: go.mod go.sum
+	@OPENAPI_PATH=`go list -m -f '{{.Dir}}' k8s.io/kube-openapi`; \
+	go run $${OPENAPI_PATH}/cmd/openapi-gen/openapi-gen.go --logtostderr \
```
Comment (zroubalik) [inline Makefile:181]: the `OPENAPI_PATH` is not resolved, as we can see in the error `stat /cmd/openapi-gen/openapi-gen.go: no such file or directory` in https://github.com/kedacore/keda/pull/1512/checks?check_run_id=1734447773#step:7:11

I had some problems with GO_PATH in the GH Action before, what was working for me locally behaved differently in GH Action :(

Comment (surki) [reply Makefile:181]: I narrowed down the problem to `go mod vendor` we are doing in Makefile and not cleaning it up after. I fixed the path resolution problem even when vendor dir is present, but we might want to stop doing the `go mod vendor` thing (or cleanup after that if really required).

Comment (zroubalik) [reply Makefile:181]: Awesome, thanks a lot for that. I do recall that the `go mod vendor` was needed to get correct clientset generation. But the vendor cleanup seems like a good idea. Thanks!

**Review by zroubalik — APPROVED** [review]
LGTM

**Pull request merged**